unsafe/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -173,18 +173,35 @@ public BytesToBytesMap(
   public Iterator<Location> iterator() {
     return new Iterator<Location>() {
 
-      private int nextPos = bitset.nextSetBit(0);
+      private int cur = 0;
+
+      private int pageCur = 0;
+
+      private MemoryBlock currentPage;
+
+      private long addr;
 
       @Override
       public boolean hasNext() {
-        return nextPos != -1;
+        return cur != size;
       }
 
       @Override
       public Location next() {
-        final int pos = nextPos;
-        nextPos = bitset.nextSetBit(nextPos + 1);
-        return loc.with(pos, 0, true);
+        if (currentPage == null) {
+          currentPage = dataPages.get(pageCur++);
+          addr = currentPage.getBaseOffset();
+        }
+        long keySize = PlatformDependent.UNSAFE.getLong(memoryManager.getPage(addr), addr);
```

Comment:
I think you're calling `getPage()` with the wrong units: `getPage` expects to receive an address that was encoded by the task memory manager, but in the previous `if` case you've assigned an offset within the page to `addr`.

This call isn't necessary, though, since we happen to have the `currentPage` already.